### DISCUSSION TITLE
The expected keywords were not found

### DISCUSSION THREAD

**Discussion by Junbo-Zheng** [BurntSushi/ripgrep#1855] [Q&A] [answered]
I had a strange experience, I want to find the expected keywords through rg command. Unfortunately, no matter how I tried, I couldn’t find it. I tried to uninstall and install again for rg package (VERSION 12.1.0 (rev e8822ce97a) now). 

However, when I try to use the ag command or grep command to replace the rg command, the keywords I expect can be found.
The screenshot is shown below: 
![Selection_014](https://user-images.githubusercontent.com/19257159/115380174-e5913600-a204-11eb-9c78-7f0e904ead01.png)
I don’t know if anyone else has encountered a similar problem, or if there is a problem with my method, if so, please correct me.

**Comment by BurntSushi** [comment, marked as answer]
Try running ripgrep with `-uuu`. ripgrep is almost certainly skipping files because of its smart filtering. You likely have relevant gitignore rules in play here.

If you want to get to bottom of it, run ripgrep with the `--debug` flag. It will tell you which files it's skipping and why.

**Reply by Junbo-Zheng** [reply]
> Try running ripgrep with `-uuu`. ripgrep is almost certainly skipping files because of its smart filtering. You likely have relevant gitignore rules in play here.
> 
> If you want to get to bottom of it, run ripgrep with the `--debug` flag. It will tell you which files it's skipping and why.

Thanks for the quick response! When I use the `-uuu` option, it works normally. Thank you!

But I still have a doubt, I tried to use `--debug` to check some logs, and there was some log output on console. However, as you know, a file like `xxx.c` can not be added to the gitignore file, because that file is a very important source file in my project, and I used `ag` and `grep` command that the expected keywords were be found. I know that `rg` has its own filtering rules, but I still feel that the keyword I expected should be found via rg without `-uuu`.

Of course, thanks for your patience, at least I know how I should use it to find the keywords I expect.

**Reply by BurntSushi** [reply]
I can't help you if you don't show me anything. :-) Did `--debug` mention the `xxx.c` file? If so, what did it say? Why did it filter it out? Just because it "can not be added to the gitignore file" doesn't mean it actually isn't.

The comparison with `ag` here doesn't really help matters. The problem with `ag` is that its gitignore support is badly broken. It often just ignores your rules. So when ripgrep respects them, it might seem surprising. I promise you, these sorts of cases are almost always gitignore configuration errors.

`grep` isn't relevant here. If `rg -uuu` finds it and `grep` finds it, then that sounds like all things are nominal. This is a question about smart filtering, which grep doesn't do, so it's no surprise that it's finding it.

Let me repeat the salient point: you must scrutinize why `xxx.c` is being ignored by looking at the `--debug` output of ripgrep. Either `xxx.c` will be mentioned explicitly or one of its parent directories will be. ripgrep should tell you which gitignore rule and file caused it to be skipped.

**Reply by BurntSushi** [reply]
If you need help reading the logs, then please post them here and I can look for you. Tell me which file you expect to be matched. Also, please don't post a screenshot it. Post it in its textual form.

**Reply by Junbo-Zheng** [reply]
I checked my project file, I think I found the cause of the problem, it is the problem of `.gitignore` file. I will show you the gitignore file:
```
.depend
Make.dep
*.o
*.a
*.d
*.i
*~
.swp
.*.swp
core
/*/
/.context
/.config
/*.lib
.kconfig
/Kconfig
```
That's the problem for `/*/`, the parent directory of the file I was expecting was filtered out, so it must not be found. Now I feel dumb, thank you for your tolerance, thanks for your support, I'm soooooo sorry!. The ripgrep is a very great tool.

**Reply by BurntSushi** [reply]
Glad you got it figured out!
